bpf_pcap.c

```diff
@@ -0,0 +1,103 @@
+#include <pcap.h>
+#include <linux/filter.h>
+#include <pcap/dlt.h>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+#include <net/if.h>
+
+#include "error.h"
+ 
+
+error_status_t dev_get_iftype(const char* if_name, int* link_type) {
+    error_status_t ret_status = STATUS_SUCCESS;
+    int sock = -1;
+    struct ifreq ifr = {0};
+    size_t if_name_len = 0;
+
+    CHECK(if_name != NULL);
+    CHECK(link_type != NULL);
+
+    sock = socket(AF_INET, SOCK_DGRAM, 0);
+    CHECK(sock >= 0);
+
+    if_name_len = strlen(if_name);
+    
+    CHECK(if_name_len < IFNAMSIZ);
+    memcpy(ifr.ifr_name, if_name, if_name_len);
+
+    // ioctl for getting hardware address
+    CHECK(ioctl(sock, SIOCGIFHWADDR, &ifr) != -1);
+
+    *link_type = ifr.ifr_hwaddr.sa_family;
+
+cleanup:
+    close(sock); // Best effort.
+    return ret_status;
+}
+
+error_status_t compile_bpf(int snaplen, int link_type, struct sock_fprog* bpf_filter, const char* bpf_str, int opt) {
+    // snaplen is  the maximum amount of data to be captured
+    // linkytpe is the link layer type? DLT_EN10MB is for ethernet, there is also DLT_NULL for BSD
+    // loopback encapsulation.
+    // netmask - is the IPv4 netmask of the network, for checking Ipv4 broadcast packets.
+    // tcpdump -ddd passes 0 to netmask so i feel comfortable doing it too.
+    error_status_t ret_status = STATUS_SUCCESS;
+    struct bpf_program _bpf = {0};
+    const struct bpf_insn* ins = NULL;
+    struct sock_filter* out = NULL;
+    int i = 0;
```

Comment:
but because in the forloop I have a couple of initializers it tries to convert all of them to `int` if i do
`for (int i = 0, ins = _bpf.bp_insns, out = bpf_filter->filter...`